share/search/index_strategy/trovesearch_denorm.py

```diff
@@ -0,0 +1,985 @@
+            return [
+                ts.propertypath_as_keyword(_path)
+                for _path in walk.paths_walked
+            ]
+
+        def _iris_by_propertypath(self, walk: ts.GraphWalk):
+            return {
+                _path_field_name(_path): ts.suffuniq_iris(ts.iris_synonyms(_iris, self.rdfdoc))
+                for _path, _iris in walk.iri_values.items()
+            }
+
+        def _iris_by_depth(self, walk: ts.GraphWalk):
+            _by_depth: dict[int, set[str]] = defaultdict(set)
+            for _path, _iris in walk.iri_values.items():
+                _by_depth[len(_path)].update(_iris)
+            return {
+                _depth_field_name(_depth): ts.suffuniq_iris(ts.iris_synonyms(_iris, self.rdfdoc))
+                for _depth, _iris in _by_depth.items()
+            }
+
+        def _texts_by_propertypath(self, walk: ts.GraphWalk):
+            return {
+                _path_field_name(_path): list(_value_set)
+                for _path, _value_set in walk.text_values.items()
+            }
+
+        def _texts_at_properties(self, walk: ts.GraphWalk, properties: Iterable[str]):
+            for _property in properties:
+                yield from walk.text_values.get((_property,), [])
+
+        def _texts_by_depth(self, walk: ts.GraphWalk):
+            _by_depth: dict[int, set[str]] = defaultdict(set)
+            for _path, _value_set in walk.text_values.items():
+                _by_depth[len(_path)].update(_value_set)
+            return {
+                _depth_field_name(_depth): list(_value_set)
+                for _depth, _value_set in _by_depth.items()
+            }
+
+        def _dates_by_propertypath(self, walk: ts.GraphWalk):
+            return {
+                _path_field_name(_path): [
+                    _date.isoformat()
+                    for _date in _value_set
+                ]
+                for _path, _value_set in walk.date_values.items()
+            }
+
+        def _ints_by_propertypath(self, walk: ts.GraphWalk):
+            return {
+                _path_field_name(_path): list(_value_set)
+                for _path, _value_set in walk.integer_values.items()
+            }
+
+        def _exact_and_suffuniq_iris(self, iri: str):
+            _synonyms = ts.iri_synonyms(iri, self.rdfdoc)
+            return {
+                'exact': list(_synonyms),
+                'suffuniq': ts.suffuniq_iris(_synonyms),
+            }
+
+    ###
+    # normalizing search responses
+
+    def _valuesearch_response(
+        self,
+        valuesearch_params: ValuesearchParams,
+        es8_response: dict,
+        cursor: _SimpleCursor,
+    ) -> ValuesearchResponse:
+        _iri_aggs = es8_response['aggregations'].get('agg_valuesearch_iris')
+        if _iri_aggs:
+            _buckets = _iri_aggs['buckets']
+            _bucket_count = len(_buckets)
+            # WARNING: terribly inefficient pagination (part two)
```

Comment:
Any guesses as to how terribly-inefficient -- worth benchmarking? It looks fairly similar to the `FlatsIndexStrat`'s pagination

Reply:
it is indeed same pagination as in `flats` -- "terribly inefficient" was a label i applied while coding (and copy-pasted here), but in practice it's not _too_ much slower to load N pages and discard N-1 of them (since N is small in practice)

maybe "terribly hacky" is a better descriptor